File: app/src/main/java/de/logerbyte/moneyminder/cashsummary/CashAdapter.java
package de.logerbyte.moneyminder.cashsummary;

import android.databinding.DataBindingUtil;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import java.util.ArrayList;

import de.logerbyte.moneyminder.R;
import de.logerbyte.moneyminder.databinding.AdapterEntryBinding;

/**
 * Created by logerom on 28.07.18.
 */

public class CashAdapter extends RecyclerView.Adapter<CashAdapter.ViewHolder> {


    private ArrayList<CashItem> list = new ArrayList<>();
    private LayoutInflater layoutInflater;

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (layoutInflater == null) {
            layoutInflater = LayoutInflater.from(parent.getContext());
        }
        AdapterEntryBinding binding = DataBindingUtil.inflate(layoutInflater, R.layout.adapter_entry, parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        CashItem cashItem = list.get(position);
        holder.binding.setViewModel(cashItem);
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    public void setList(ArrayList<CashItem> list) {
        this.list = list;
    }

    // TODO: 29.07.18 why an extra viewHolder?
    protected static class ViewHolder extends RecyclerView.ViewHolder {

        AdapterEntryBinding binding;

        public ViewHolder(AdapterEntryBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/main/java/de/logerbyte/moneyminder/presentation/cashadapter/CashAdapter.kt
package de.logerbyte.moneyminder.presentation.cashadapter

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.RecyclerView
import de.logerbyte.moneyminder.R
import de.logerbyte.moneyminder.data.viewItem.DayExpenseViewItem
import de.logerbyte.moneyminder.data.viewItem.DayExpenseViewViewItem
import de.logerbyte.moneyminder.data.viewItem.ExpenseListViewItem
import de.logerbyte.moneyminder.data.viewItem.ExpenseListViewType
import de.logerbyte.moneyminder.data.viewItem.SummaryMonthViewItem
import de.logerbyte.moneyminder.presentation.dialog.dialogDelete.DeleteDialogFragment
import de.logerbyte.moneyminder.domain.database.expense.Expense
import de.logerbyte.moneyminder.databinding.AdapterEntryBinding
import de.logerbyte.moneyminder.databinding.AdapterEntryPlusSummaryBinding
import de.logerbyte.moneyminder.domain.ExpenseDataManager
import java.text.SimpleDateFormat
import java.util.*
import javax.inject.Inject
import kotlin.collections.ArrayList
import kotlin.collections.HashMap

/**
 * Created by logerom on 28.07.18.
 */

class CashAdapter(
    val onDeleteClicked: () -> View) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private lateinit var layoutInflater: LayoutInflater
    var items = listOf<ExpenseListViewItem>()
        set(value) {
            field = value
            this.notifyDataSetChanged()
        }

    override fun getItemCount(): Int {
        return items.size
    }

    override fun getItemViewType(position: Int): Int {
        return when(items[position]){
            is DayExpenseViewItem -> ExpenseListViewType.DAY_ITEM.ordinal
            is SummaryMonthViewItem -> ExpenseListViewType.MONTH_ITEM.ordinal
            else -> -1
        }
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        layoutInflater = LayoutInflater.from(recyclerView.context)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if(viewType == ExpenseListViewType.DAY_ITEM.ordinal)
            DayExpenseViewHolder(AdapterEntryBinding.inflate(layoutInflater))
        else MonthExpenseViewHolder(AdapterEntryPlusSummaryBinding.inflate(layoutInflater))
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (holder){
            is DayExpenseViewHolder -> {holder.bind(items[position] as DayExpenseViewItem)}
            is MonthExpenseViewHolder -> holder.bind(items[position] as SummaryMonthViewItem)
        }
    }

    inner class DayExpenseViewHolder(val binding: AdapterEntryBinding): RecyclerView.ViewHolder(binding.root) {
        fun bind(viewItem: DayExpenseViewItem){
            binding.vmCashItem = viewItem
            binding.buDelete.setOnClickListener { onDeleteClicked }
        }
    }

    inner class MonthExpenseViewHolder(val binding: AdapterEntryPlusSummaryBinding): RecyclerView.ViewHolder(binding.root) {
        fun bind(viewItem: SummaryMonthViewItem){
            binding.viewItem = viewItem
        }
    }
}
